### ISSUE TITLE
Firestore onSnapshot() will silently fail when forbidden by security rules

### ISSUE THREAD

**Issue by icoloma** [firebase/firebase-js-sdk#5856]
* Operating System version: Ubuntu 20.04.3 LTS
  * Browser version: Chrome Version 95.0.4638.69 (Official Build) (64-bit)
  * Firebase SDK version: 9.6.1
  * Firebase Product: Firestore

#### Describe the issue

When using `onSnapshot()` with a docRef that is forbidden to be read by the security rules, it fails silently. No error is logged in the browser console or the Firestore emulator. To find out what happened, I need to use `getDoc()` instead.

#### Steps to reproduce:

firestore.rules
```
rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write: if false;
    }
  }
}
```

Relevant code:

```javascript
const ref = doc(db, `foo/id`);
// use this to see an error message in the browser console
// console.log((await getDoc(ref)).data);

// this fails silently
const unsubscribe = onSnapshot(ref,
    (snapshot) => {
       console.log("Snapshot received", snapshot);
    },
    error => {
       console.error(error);
   }
);
```

#### What should happen

I should get an error on the browser console or the Emulator log.

**Comment by dconeybe** [comment]
Hi @icoloma. Thanks for reporting this issue. I'll investigate and get back to you.

**Comment by dconeybe** [comment]
@icoloma I am not able to reproduce the bug. When I run your code the `console.error(error)` line is invoked, as expected.

Here is the code that I used to attempt to reproduce: https://github.com/dconeybe/SnapshotListenerSecurityRulesDenyIssue5856

Here is the output that I got, which shows that the error callback is correctly invoked:

```
onSnapshot error: FirestoreError [FirebaseError]: Missing or insufficient permissions.
    at fromRpcStatus (node_modules/@firebase/firestore/src/remote/serializer.ts:153:10)
    at fromWatchChange (node_modules/@firebase/firestore/src/remote/serializer.ts:477:33)
    at PersistentListenStream.onMessage (node_modules/@firebase/firestore/src/remote/persistent_stream.ts:642:25)
    at node_modules/@firebase/firestore/src/remote/persistent_stream.ts:517:21
    at node_modules/@firebase/firestore/src/remote/persistent_stream.ts:570:18
    at node_modules/@firebase/firestore/src/util/async_queue_impl.ts:135:7
    at node_modules/@firebase/firestore/src/util/async_queue_impl.ts:186:14
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  code: 'permission-denied',
  toString: [Function (anonymous)]
}
```

Can you try to reproduce with my test app and see what behavior you get?

**Comment by icoloma** [comment]
I'm awfully sorry. I thought I tested it thoroughly, but apparently I missed something in between. Your test works perfectly fine in my local environment, so I tried to reproduce again in my own app and couldn't. Apparently I have been blindsided somewhere along the way (or something was outdated while upgrading Firebase). 

I'm closing the issue, sorry for consuming your time with this, and huge thanks for the repro case.
